### PR TITLE
Add engine_process SimInspector

### PR DESCRIPTION
**Pull request by tyb0807**
Hello,

In answer to the request #529, here's my first attempt. Please tell me if I need to fix/add anything.

Thanks @zardus and @rhelmot.

### CODE REVIEW

**Comment by humeafo** [comment]
I'm not very familiar with angr codes, but some points to discuss:
1. is the arg-name sim_engine/sim_successors is a little better and conforming to angr naming conventions?
2. state._inspect('engine_process', when=BP_AFTER, simsuccessor=successors)
should be new_state._inspect(...) after engine processing?
3. after _inspect call, should the successors to be re-assigned to the returned successors if inspect user changed successors?

**Comment by tyb0807** [comment]
Hello,

1. I followed attribute naming as in `simprocedure` event_type, but yes, I prefer `sim_engine` and `sim_successor` as well.
2. `new_state` after `_process` function will really be one of the successors of `state`, so I think it would be more useful to let people access the old state, because we'll have the new state in `successors` anyway.
3. I had exactly a conversation on this with @zardus, personally I'm not sure if we should let users change successors, but feel free to let me know if you have that needs.

I'm not that familiar with angr code either, so I'm just trying to learn, I'll let @rhelmot @ltfish @zardus decide anyways.

Thank you

**Comment by zardus** [comment]
1. I agree with `sim_engine` and `sim_successors`. Just cause we were noobs at naming stuff before, doesn't mean we have to keep screwing up :-)
2. I agree that we should trigger the BP_AFTER with the _old_ state. That makes more sense.
3. I think we _should_ support SimSuccessors modification. Let's err on the side of flexibility.

Otherwise, awesome stuff!

**Comment by tyb0807** [comment]
OK, I've modified things. Please tell me if you need anything else.

File: angr/engines/engine.py [hunk 1]
```diff
@@ -41,9 +41,11 @@ def process(self, state, *args, **kwargs):
 
         successors = SimSuccessors(addr, state)
 
-        state._inspect('engine_process', when=BP_BEFORE, simengine=self)
+        state._inspect('engine_process', when=BP_BEFORE, sim_engine=self, sim_successor=successors)
+        successors = state._inspect_getattr('sim_successor', successors)
         self._process(new_state, successors, *args, **kwargs)
-        state._inspect('engine_process', when=BP_AFTER, simsuccessor=successors)
+        state._inspect('engine_process', when=BP_AFTER, sim_successor=successors)
```
Comment (humeafo) [inline angr/engines/engine.py]: sim_successor should be sim_successors?

**Comment by tyb0807** [comment]
Definitely, don't know why I always thought that `SimSuccessors` class is `SimSuccessor`...

Fixed. Thanks,

**Comment by humeafo** [comment]
This was not merged yet, was it refused or just forgotten?

**Comment by zardus** [comment]
It looks like github auto-closed it when the `wip/the_end_times` branch was deleted... I merged it into master now! Rebased into commit https://github.com/angr/angr/commit/58d0120f4a56e604c6445bb4da5f7bea833c5c90.

**Pull request closed without merging**